program/src/state.rs

```diff
@@ -3,58 +3,546 @@
 
 //! State transition types
 
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::marker::PhantomData;
 use std::ops::Range;
 
 use serde::Serialize;
 
 use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
-use solana_program::borsh::{get_instance_packed_len, try_from_slice_unchecked};
-use solana_program::clock::Clock;
 use solana_program::{
-    account_info::AccountInfo, clock::Epoch, entrypoint::ProgramResult, msg,
-    program_error::ProgramError, program_pack::Pack, pubkey::Pubkey, rent::Rent, sysvar::Sysvar,
+    account_info::AccountInfo,
+    borsh::try_from_slice_unchecked,
+    clock::Clock,
+    clock::Epoch,
+    entrypoint::ProgramResult,
+    msg,
+    program_error::ProgramError,
+    program_memory::sol_memcmp,
+    program_pack::Pack,
+    program_pack::Sealed,
+    pubkey::{Pubkey, PUBKEY_BYTES},
+    rent::Rent,
+    sysvar::Sysvar,
 };
 use spl_token::state::Mint;
 
+use crate::big_vec::BigVec;
 use crate::error::LidoError;
 use crate::logic::get_reserve_available_balance;
 use crate::metrics::Metrics;
 use crate::processor::StakeType;
 use crate::token::{self, Lamports, Rational, StLamports};
 use crate::util::serialize_b58;
 use crate::{
-    account_map::{AccountMap, AccountSet, EntryConstantSize, PubkeyAndEntry},
     MINIMUM_STAKE_ACCOUNT_BALANCE, MINT_AUTHORITY, RESERVE_ACCOUNT, STAKE_AUTHORITY,
+    VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT,
 };
-use crate::{VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT};
 
 pub const LIDO_VERSION: u8 = 1;
 
 /// Size of a serialized `Lido` struct excluding validators and maintainers.
 ///
 /// To update this, run the tests and replace the value here with the test output.
 pub const LIDO_CONSTANT_SIZE: usize = 353;
-pub const VALIDATOR_CONSTANT_SIZE: usize = 49;
 
-pub type Validators = AccountMap<Validator>;
+/// Types of list entries
+/// Uninitialized should always be a first enum field as it catches empty list data errors
+#[derive(Clone, Debug, PartialEq, Eq, BorshDeserialize, BorshSerialize, Serialize, BorshSchema)]
+pub enum ListEntryType {
+    /// If the account has not been initialized, the enum will be 0
+    Uninitialized,
+    Validator,
+    Maintainer,
+}
```

Comment:
To know a type of serialized list elements. Otherwise when deserializing there is now way to know a type of list elements. Although we could use a knowledge that first goes Lido, then maintainer list and then validator list in raw data `[Lido, validators, maintainers]` but it is error prone due to forgetting that and swapping the order. So the list type is always checked.
Also to produce generic messages: ` msg!("Pubkey {} is duplicated in a {:?} list", pubkey, T::TYPE)`
Also to choose an algorithm as in `fn deserialize_account_list<T: ListEntry>`

Reply:
> but it is error prone due to forgetting that and swapping the order.

I would argue it’s the other way around, by introducing this, you introduce new representable states that are invalid. A big part of leveraging type systems to eliminate bugs is to make invalid states unrepresentable. You *still* depend on the order with the tag, but now you also need to verify that the tag matches the expectation, and add error handling logic for when it does not.